Add training mode

robintw opened the pull request:
## 🧰 Issue
Closes #614 

## 🚀 Overview: 
Adds a `--training` parameter to both Pepys Import and Pepys Admin, which uses a temporary database in the user's home folder - so the user can experiment with Pepys without affecting the real live database. Includes a new Start Menu shortcut for 'Pepys Admin (training mode)' and a new Send To shortcut for 'Pepys Import (training mode)'

## 🔗 Link to preview (or screenshot, if relevant)
![screenshot1](https://user-images.githubusercontent.com/296686/99544914-bd2eb280-29ac-11eb-88c5-651727be149d.png)
<img width="691" alt="Screen Shot 2020-11-18 at 14 46 07" src="https://user-images.githubusercontent.com/296686/99544995-d46da000-29ac-11eb-942c-d6333e38f3d9.png">

## 🤔 Reason: 
Allows the user to do the Pepys training without affecting any real databases.

## 🔨Work carried out:
- [x] Manual tests on Windows show that Send To and Start Menu shortcut work
- [x] Write automated tests for `--training` parameter
- [ ] Tests pass


## Confirmations

- [ ] I have chosen reviewers for my PR.
- [x] I have chosen an appropriate label for the PR.
- [x] I have completed the mandatory sections of this document.
- [x] I have deleted any unused sections.

## 📝 Developer Notes:
A few of the changes to do with the tests for `--training` are below:

We had to deal with an issue when using --training in tests, as changes to environment variables stick around for the whole run of the process. This isn't a problem for normal use, as we start a process to run Pepys (with or without --training), run it, and then finish the process. If we have to change the PEPYS_CONFIG_FILE env var to deal with --training, then it only takes effect for that process.

However, if we're running in Pytest, we have a single process that keeps going for all the tests, and so the changed env var (and config resulting from it) sticks around.

The fix is to reset the PEPYS_CONFIG_FILE env var at the end of the --training tests, and also change the cli.py files (for Pepys Admin and Pepys Import) to reload the config whenever they are run, in case the PEPYS_CONFIG_FILE env var has changed.

robintw commented:
This is completed and working, except that we haven't yet implemented loading of some sample data into the training database. There was some discussion on Slack about this. @IanMayo @clenoble: what do you want me to do about this? I can see the advantages of both options: having pre-loaded data gives users a more realistic experience, but having a blank database makes it clearer what users have added (and saves us importing data that they might want to import again later, which may confuse them).

Also, once #625 is merged the tests will pass for this - it's a benchmark test failure.

clenoble commented:
@robintw @IanMayo I think keeping it as close to reality as possible is best. In practice, analysts will check with Pepys Viewer or directly in Debrief or ArcGIS (Pepys Viewer will tell them some data have been added, but they won't know if it's the right data, or if Pepys interpreted it correctly).
Is it possible to re-initialize the training database each time a training session starts? So if several users do the training on the same terminal, they can all import the same .gpx file again and again? Or a user wants to repeat the training after not using the tool for a few months or so.

IanMayo commented:
> re-initialize data

If we're using training mode within our tutorial, users will probably run Pepys-Import (or viewer) several times. For this, I think we need the data to be persistent (and not wipe it each time).  If we're going to demonstrate how Pepys-Import handles a user trying to import a datafile twice, then we need the data to be persistent between sessions.

> repeat the training

Hmm, yes - I guess our training mode needs a "Re-initialise database" command.  Maybe when it first opens?  i.e. If Pepys-Import finds an existing training database it offers to re-initialise it. If they say yes, delete the database, and move back to the "database not found - create training database" step.

Our plan is to store the training database in the user's home directory - this will prevent multiple users seeing the same database.

I'll create a separate issue covering the provision of default data in the training database, since we may want to put some thought into what training data to provide.

IanMayo commented:
@robintw - now that I've moved Training Data to a separate issue, I think that "re-initialise database" is the only outstanding feature for this PR.

robintw commented:
Thanks @IanMayo.

I'm wondering whether it's a good idea to add that question to Pepys Import/Pepys Admin, as it's an extra UI step that won't be there when they're running Pepys 'for real'. What do you think about adding another Start Menu shortcut called something like "Reinitialise Training Mode" or "Reset Training Mode" or something? @clenoble - any thoughts?

clenoble commented:
@robintw @IanMayo Having a separate way (through start menu) for resetting the training mode will be very confusing for users who are used to access the training mode via "Send to". 
Having a separate access for pepys-admin is okay because only expert users will need and use it. For most users and especially users who need training, I think we should keep the "Send to" access and suggest to the user to reset the database when they start, and also once a file has been uploaded.

IanMayo commented:
Thanks @clenoble 

>  and also once a file has been uploaded.
So, they right click on a file, select `Send to > Pepys (training mode)`, answer the questions to load the data` then Pepys asks them if they want to reset the database?


I think the main part of what you're asking for could be provided in this way:
* when Pepys-Import starts in `--training` mode, if it finds an existing training database, it asks the user if they want the database to be re-initialised.  If the user does, the database is deleted, and Pepys goes through the process of creating (and populating) a training database.

clenoble commented:
@IanMayo yes, I agree with giving the possibility to reset when Pepys-Import starts, but I would like it to also be an option at the end. Otherwise, the user who has just finished the training and want to clear the database has to fake sending a file to Pepys training mode to be able to reset the database and then cancel the import. It is a weird process.

IanMayo commented:
Ok, thanks @clenoble .  I guess the equivalent is to offer to delete the database before exiting Pepys-Import. Then they could handle a whole folder of files, and just delete their database before quitting.  For the next time they enter training mode, it will create a fresh database.

Would that be an acceptable solution?

clenoble commented:
@IanMayo yes, that works well.

robintw commented:
@IanMayo @clenoble: I've added a question at the beginning and the end of the Pepys Import/Pepys Admin run asking whether to reset the database. It looks like this:

<img width="752" alt="Screen Shot 2020-11-25 at 09 43 00" src="https://user-images.githubusercontent.com/296686/100210438-e7c7c080-2f02-11eb-9c8f-08c84bd7f81b.png">

I've also added a demo notebook and labelled this issue so that Binder builds automatically. Once the Binder link appears in a comment here, click that and then choose 'Demo Training Mode' and you can have a play to check it works how you think it should.

IanMayo commented:
I'm playing with the Jupyter preview - it's great :-)

I _think_ we could do with more explanation on training mode.

Note: if there is an existing database, the `Running in training mode` is in a rectangular table.  If there isn't one, it's a plain statement.

I _think_ it should be in rectangular border in both cases.  Plus, I think we should explain what training mode is, so:
```
##############################################
#          Running in training mode          #
#                                            #
# Changes are only made to database in:      #
# /home/jovyan/pepys_training_database.db    #
#                                            #
##############################################

```

IanMayo reviewed (approved):
Works fine.

_Hopefully_ the commits with the line-ending changes will get rid of a local issue I have with MACOS mangling the endings.

BarisSari reviewed (approved):
It looks great, thanks @robintw. I'm approving the PR but the only thing that concerns me is that if e_trac and e_trac_bad files use CRLF line endings or not. Because it becomes painful to work with that in the Mac environment.

robintw commented on tests/sample_data/track_files/other_data/e_trac.txt:
```diff
@@ -1,45 +1,45 @@
+4,143732318,  2019/08/06,04:44:00,  -26.1706666, 43.33,0,0,511,  -128,0,0,15,  -1,  -1,  -1,56,16343, 16322 DINL
+4,143732286,  2019/08/06,04:40:00,  -26.8973333, 41.9753333,13,289,511,  -128,0,0,15,  -1,  -1,  -1,56,16344, 16323 MEGH
+4,143732345,  2019/08/06,04:40:00,  -28.0873333, 42.1813333,9,12,0,  -128,0,0,15,  -1,  -1,  -1,56,16345, 16316 PELA
+4,143732319,  2019/08/06,04:48:00,  -26.674, 43.2753333,0,98,511,  -128,0,0,15,  -1,  -1,  -1,56,16346, 16331 FLIS
```
I've checked and they are definitely LF in this branch.

robintw commented on tests/sample_data/track_files/other_data/e_trac_bad.txt:
```diff
@@ -1,46 +1,46 @@
+4,143732345,  2019/08/06,04:40,  -28.0873333, 42.1813333,9,12,0,  -128,0,0,15,  -1,  -1,  -1,56,16345, 16316 PELA
+4,143732319,  2019/08/06,04:48:00,  -26.674, 43.2753333,0,98,511,  -128,0,0,15,  -1,  -1,  -1,56,16346, 16331 FLIS
+4,143732307,  2019/08/06,04:48:00,  -28.6573333, 42.1713333,7,283,511
+
```
I've checked and they are definitely LF in this branch.
